File: webrtc-sys/src/frame_transform.cpp
/*
 * VeilMesh patch (D6): see include/livekit/frame_transform.h.
 * Licensed under the Apache License, Version 2.0.
 */
#include "livekit/frame_transform.h"

#include "api/make_ref_counted.h"
#include "webrtc-sys/src/frame_transform.rs.h"

namespace livekit_ffi {

RustFrameTransformer::RustFrameTransformer(
    rust::Box<FrameTransformHandler> handler)
    : handler_(std::move(handler)) {}

RustFrameTransformer::~RustFrameTransformer() = default;

void RustFrameTransformer::Transform(
    std::unique_ptr<webrtc::TransformableFrameInterface> frame) {
  const bool outgoing =
      frame->GetDirection() ==
      webrtc::TransformableFrameInterface::Direction::kSender;
  const bool is_video = frame->GetMimeType().rfind("video/", 0) == 0;
  const bool key_frame =
      is_video &&
      static_cast<webrtc::TransformableVideoFrameInterface*>(frame.get())
          ->IsKeyFrame();

  auto data = frame->GetData();
  rust::Slice<const uint8_t> input(data.data(), data.size());
  rust::Vec<uint8_t> output = handler_->transform(
      outgoing, is_video, key_frame, frame->GetSsrc(), frame->GetTimestamp(),
      input);
  if (output.empty()) {
    // Rust dropped the frame (e.g. undecryptable): nothing goes further.
    return;
  }
  frame->SetData(std::span<const uint8_t>(output.data(), output.size()));

  webrtc::scoped_refptr<webrtc::TransformedFrameCallback> callback;
  {
    std::lock_guard<std::mutex> lock(mutex_);
    auto it = sink_callbacks_.find(frame->GetSsrc());
    callback = it != sink_callbacks_.end() ? it->second : callback_;
  }
  if (callback) {
    callback->OnTransformedFrame(std::move(frame));
  }
}

void RustFrameTransformer::RegisterTransformedFrameCallback(
    webrtc::scoped_refptr<webrtc::TransformedFrameCallback> callback) {
  std::lock_guard<std::mutex> lock(mutex_);
  callback_ = std::move(callback);
}

void RustFrameTransformer::RegisterTransformedFrameSinkCallback(
    webrtc::scoped_refptr<webrtc::TransformedFrameCallback> callback,
    uint32_t ssrc) {
  std::lock_guard<std::mutex> lock(mutex_);
  sink_callbacks_[ssrc] = std::move(callback);
}

void RustFrameTransformer::UnregisterTransformedFrameCallback() {
  std::lock_guard<std::mutex> lock(mutex_);
  callback_ = nullptr;
}

void RustFrameTransformer::UnregisterTransformedFrameSinkCallback(
    uint32_t ssrc) {
  std::lock_guard<std::mutex> lock(mutex_);
  sink_callbacks_.erase(ssrc);
}

void set_sender_frame_transform(std::shared_ptr<RtpSender> sender,
                                rust::Box<FrameTransformHandler> handler) {
  auto transformer =
      webrtc::make_ref_counted<RustFrameTransformer>(std::move(handler));
  sender->rtc_sender()->SetEncoderToPacketizerFrameTransformer(transformer);
}

void set_receiver_frame_transform(std::shared_ptr<RtpReceiver> receiver,
                                  rust::Box<FrameTransformHandler> handler) {
  auto transformer =
      webrtc::make_ref_counted<RustFrameTransformer>(std::move(handler));
  receiver->rtc_receiver()->SetDepacketizerToDecoderFrameTransformer(
      transformer);
}

void clear_sender_frame_transform(std::shared_ptr<RtpSender> sender) {
  sender->rtc_sender()->SetEncoderToPacketizerFrameTransformer(nullptr);
}

void clear_receiver_frame_transform(std::shared_ptr<RtpReceiver> receiver) {
  receiver->rtc_receiver()->SetDepacketizerToDecoderFrameTransformer(nullptr);
}

}  // namespace livekit_ffi

File: webrtc-sys/src/frame_transform.rs
// VeilMesh patch (D6): frame transform hook — every encoded frame passes
// through Rust. See include/livekit/frame_transform.h.
// Licensed under the Apache License, Version 2.0.

use std::sync::Arc;

#[cxx::bridge(namespace = "livekit_ffi")]
pub mod ffi {
    unsafe extern "C++" {
        include!("livekit/frame_transform.h");
        include!("livekit/rtp_sender.h");
        include!("livekit/rtp_receiver.h");

        type RtpSender = crate::rtp_sender::ffi::RtpSender;
        type RtpReceiver = crate::rtp_receiver::ffi::RtpReceiver;

        pub fn set_sender_frame_transform(
            sender: SharedPtr<RtpSender>,
            handler: Box<FrameTransformHandler>,
        );
        pub fn set_receiver_frame_transform(
            receiver: SharedPtr<RtpReceiver>,
            handler: Box<FrameTransformHandler>,
        );
        pub fn clear_sender_frame_transform(sender: SharedPtr<RtpSender>);
        pub fn clear_receiver_frame_transform(receiver: SharedPtr<RtpReceiver>);
    }

    extern "Rust" {
        type FrameTransformHandler;

        /// `outgoing` — encoder→packetizer (encrypt), otherwise
        /// depacketizer→decoder (decrypt). Empty result drops the frame.
        fn transform(
            self: &FrameTransformHandler,
            outgoing: bool,
            is_video: bool,
            key_frame: bool,
            ssrc: u32,
            rtp_timestamp: u32,
            data: &[u8],
        ) -> Vec<u8>;
    }
}

/// What Rust does with a frame. Called on libwebrtc's encoder/decoder
/// threads: must be cheap and must not block on the signaling thread.
pub trait FrameTransform: Send + Sync {
    fn transform(
        &self,
        outgoing: bool,
        is_video: bool,
        key_frame: bool,
        ssrc: u32,
        rtp_timestamp: u32,
        data: &[u8],
    ) -> Vec<u8>;
}

pub struct FrameTransformHandler {
    inner: Arc<dyn FrameTransform>,
}

impl FrameTransformHandler {
    pub fn new(inner: Arc<dyn FrameTransform>) -> Self {
        Self { inner }
    }

    fn transform(
        &self,
        outgoing: bool,
        is_video: bool,
        key_frame: bool,
        ssrc: u32,
        rtp_timestamp: u32,
        data: &[u8],
    ) -> Vec<u8> {
        self.inner
            .transform(outgoing, is_video, key_frame, ssrc, rtp_timestamp, data)
    }
}
